src/java/com/puppetlabs/jruby_utils/pool/JRubyPool.java

```diff
@@ -491,8 +491,7 @@ private void signalIfLockCanProceed() {
         // is active at a time - a caller of lock() that has just acquired
         // the pool lock but is waiting for all registered elements to be
         // returned to the queue.
-        if (registeredElements.size() == liveQueue.size() ||
-                pill != null) {
+        if (this.maxSize == liveQueue.size() || pill != null) {
```

Comment:
We did made this change in the `lock()` method as part of #34.  I think @jpinsonault was just cleaning up this other reference in `signalIfLockCanProceed()` to do the same.

I originally had had this key off of `registeredElements` rather than `liveQueue` size because I was thinking of a future world in which we might be able to dynamically size the `maxSize` of the pool.  For that case, I had been thinking that if a lock were requested that the unregistering of an element from the pool should, assuming all other still registered instances are in the pool, allow the lock to be granted/proceed.  For #34, we changed the flush implementation to acquire a lock before borrowing all instances but, for performance, release the lock before filling the pool back up.  For the case where two flush lock requests and one or more separate borrow requests overlap one another, waiting for `registeredElements` to equal `liveQueue` size created a race condition where the second lock request could be granted before the pool was filled back up and, thereby, create a "deadlock" where the second locker would hang forever when trying to borrow an instance out of the pool - because the separate borrow request was signaled that the instance was available to be borrowed rather than the second locker.

I can think of some other ways we could have worked around this problem without giving up the ability to let the pool be dynamically sizable in the future.  We don't actually have a requirement yet for a dynamically sizable pool, though - and may never have one.  This seemed like a much simpler change to head off the "deadlock" problem than other options.

Hope some of that made sense.  @jpinsonault admirably spent a lot of time diagnosing this problem when it first came up and had us both scratching our heads for quite a while before we were able to figure out what was going on.